README.md
# Table of Content

- [Welcome to *Swarm-Rescue*](#welcome-to-swarm-rescue)
- [The Competition](#the-competition)
- [Simple-Playgrounds](#simple-playgrounds)
- [Installation on Ubuntu](#installation-on-ubuntu)
- [Installation on Windows 10](#installation-on-windows-10)
- [Elements of the environment](#elements-of-the-environment)
- [Programming](#programming)
- [Contact](#contact)

# Welcome to *Swarm-Rescue*

With this project, you will try to save human lives, in simulation... Teach a swarm of drones how to behave to save a maximum of injured people in a minimum of time!

Your job will be to propose your own version of the controller of the drone. In a competition, each participating team will perform on a new unknown map, the winner will be the one who gets the most points based on several criteria: speed, quality of exploration, number of injured people saved, number of drones returned, etc.

*Swarm-Rescue* is the environment that simulates the drones and that describes the maps used, the drones and the different elements of the map.

[Access to the GitHub repository *Swarm-Rescue*](https://github.com/emmanuel-battesti/swarm-rescue)

The Challenge does not require any particular technical skills (beyond basic knowledge of *Python*), and will mainly mobilize creativity and curiosity from the participants.

# The Competition

The objective of the mission is to map an unknown, potentially hostile area, detect targets (injured or healthy people), and guide them out of the area. A typical use case is to investigate the basement of a collapsed building in the dark, in order to locate trapped people and rescue them.

Each team will have a fleet of 10 drones. Each drone will be equipped with communication functions and sensors.

**Your job is to make these drones completely autonomous by programming them in *Python*.**

The drones will have to manage the limited range of the communication means, collaborate between them to acquire information and share information, be able to manage sensor and communication means failures and unforeseen events such as the loss of drones in order to conduct this mission autonomously.

The work on the challenge will be done **exclusively in this simulation environment**, with maps of increasing complexity. The final evaluation will be done on an unknown map designed by the organizers and not available to the contestants. Every proposition will be tested on a same computer and a score related to the performance will be computed.


## Score

Depending on the scenario evolution before the final evaluation, the score calculation may be slightly modified.

First, the execution of your algorithms will be stopped after a while. There are two time-limits:
- time step limit: a number of loops in the simulation
- real time limit: a limit in minutes, depending on the map and the computer speed: 2 to 5 minutes.

When the first limit is reached, the game is over. If your algorithm is fast, you will reach the "step time limit" first. If your algorithm is too slow, you will reach the "real time limit" before the "time step limit" .

To calculate the score, the following elements will be taken into account:
- the part of the territory explored: you have to explore a maximum of the map
- the number of fixed targets detected, brought back to the base,
- the number of mobile targets detected, brought back to the base.

If all the targets are brought back to base and all the map is explored, the speed in "time step" will be taken into account.

## Details on the rules

- You can only use the detection of WoundedPerson, RescueCenter and other Drones with the Semantic sensor, but not the Walls (as it is unrealistically giving you the full wall size and position). Use Lidar to detect/avoid Walls.

# Simple-Playgrounds 

This program *Swarm-Rescue* is an extension of the *Simple-Playgrounds* (SPG) software library: [https://github.com/mgarciaortiz/simple-playgrounds](https://github.com/mgarciaortiz/simple-playgrounds). However, in the current installation of *Swarm-Rescue*, it is the branch *swarm-rescue-v2* of a fork of *Simple-Playgrounds* that is used: [https://github.com/emmanuel-battesti/simple-playgrounds](https://github.com/emmanuel-battesti/simple-playgrounds).


It is recommended to read the [documentation of *Simple-Playgrounds*](https://github.com/emmanuel-battesti/simple-playgrounds#readme).

*Simple-Playgrounds* is an easy-to-use, fast and flexible simulation environment. It bridges the gap between simple and efficient grid environments, and complex and challenging 3D environments. It proposes a large diversity of environments for embodied drones learning through physical interactions. The playgrounds are 2D environments where drones can move around and interact with scene elements.

The game engine, based on [Pymunk](http://www.pymunk.org) and [Arcade](https://api.arcade.academy/), deals with simple physics, such as collision and friction. Drones can act through continuous movements and discrete interactive actions. They perceive the scene with realistic first-person view sensors, top-down view sensors, and semantic sensors.

## Game Engine

In *Simple-Playgrounds*, the game engine used is *Arcade*. Drones enter a Playground, and start acting and perceiving within this environment. The perception/action/communication loop is managed by the game engine. At each time step, all perception is acquired, all communication are done. Then according to actions to do, drones are moved. Everything is synchronized, unlike what you would get on a real robot.

## Physics Engine

In *Simple-Playgrounds*, the 2d physics library *Pymunk* is used. The physic engine deals with simple physics, such as collision and friction. This gives a mass and inertia to all objects.


# Installation on Ubuntu

This installation procedure has been tested with Ubuntu 18.04 and 20.04.

## Arcade library dependencies

First, you will obviously have to use the Git tool.

And for the library *Arcade*, you might need to install *libjpeg-dev* and *zlib1g-dev*.

```bash
sudo apt update
sudo apt install git libjpeg-dev zlib1g-dev
```

## *Python* installation

We need, at least, *Python 3.8*.

- On *Ubuntu 20.04*, the default version of *Python* is 3.8.
- On *Ubuntu 18.04*, the default version of *Python* is 2.7.17. And the default version of *Python3* is 3.6.9.

But it is easy to install *Python* 3.8:
```bash
sudo apt update
sudo apt install python3.8 python3.8-venv python3.8-dev
```

## *Pip* installation

- Install *Pip*:

```bash
sudo apt update
sudo apt install python3-pip 

- When the installation is complete, verify the installation by checking the *Pip* version:

```bash
pip3 --version
```

- It can be useful to upgrade *Pip* to have the last version in local directory:

```bash
/usr/bin/python3.8 -m pip install --upgrade pip
```

To use the correct version, you have to use `python3.8 -m pip` instead of `pip`, for example:

```bash
python3.8 -m pip --version
```

## Virtual environment tools

The safe way to work with *Python* is to create a virtual environment around the project.

For that, you should install some tools:

```bash
sudo apt update
sudo apt install virtualenvwrapper
```
## Install this *swarm-rescue* repository

- To install this git repository, go to the directory you want to work in (for example: *~/code/*).

- Git-clone the code of [*Swarm-Rescue*](https://github.com/emmanuel-battesti/swarm-rescue):

```bash
git clone https://github.com/emmanuel-battesti/swarm-rescue.git
```
This command will create the directory *swarm-rescue* with all the code inside it.

- Create your virtual environment. This command will create a directory *env* where all dependencies will be installed:

```bash
cd swarm-rescue
python3.8 -m venv env
```

- To use this newly create virtual environment, as each time you need it, use the command:

```bash
source env/bin/activate
```

To deactivate this virtual environment, simply type: `deactivate`

- With this virtual environment activated, we can install all the dependency with the command:

```bash
python3.8 -m pip install --upgrade pip
python3.8 -m pip install -r requirements.txt
```

- To test, you can launch:

```bash
python3.8 ./src/swarm_rescue/launcher.py
```

## Python IDE

Although not mandatory, it is a good idea to use an IDE to code in *Python*. It makes programming easier.

For example, you can use the free *community* version of [*PyCharm*](https://www.jetbrains.com/pycharm/). In this case, you have to set your *interpreter* path to your venv path to make it work. 


# Installation on Windows 10

This installation procedure has been tested with Windows 10. Installation is also straightforward on Windows 11.

## *Python* installation

- Open this link in your web browser:  https://www.python.org/downloads/windows/
- Don't choose the latest version of Python, but choose the 3.8 version. Currently (11/2022), it is the "*Python 3.8.10 - May 3, 2021*".
- For modern machine, you have to choose the *Windows x86-64 executable installer*.
- Once the installer is downloaded, run the Python installer.
- **Important** : you should check the "**Add Python 3.8 to path**"  check box to include the interpreter in the execution path.

## *Git* installation

Git is a tool for source code management. [Git is used](https://www.simplilearn.com/tutorials/git-tutorial/what-is-git "Git is used") to tracking changes in the source code of *swarm-rescue*.

 - Download the [latest version of    Git](https://git-scm.com/download/win) and choose the 64/32 bit version.
 - After the file is downloaded, install it in the system.
 - Once installed, select *Launch the Git Bash*, then click on *finish*. The *Git Bash* is now launched.

We want to work later on the project by using the *Git Bash* terminal.

## Configure *Git Bash*

- Launch the *Git Bash* terminal
- **Warning**, you are **not** by default to your home directory. So to go there, just type : *cd*
- To facilitate the use of the command *python*, you have to create an alias to real position of the program python.exe : `echo "alias python='winpty python.exe'" >> ~/.bashrc`

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [Unreleased]

## [2.2.3] - 2023-03-20

### Changed
- Change the tool image_to_map.py

### Added
- Add the map of the final !

### Fixed
- Now, round terminates when the time_step exceeds the time_step_limit

## [2.2.2] - 2023-02-07

### Changed
- Walls are no longer detected as OTHER, they are remove from the results of the semantic sensor.
- Improve the generated report. Now, in english.

### Added
- Add a new example for the semantic sensor.

### Fixed
- Drones are correctly detected as DRONE in the results of the semantic sensor.
- Bug fix concerning a random crash.
- Bug fix in the computation of the number of wounded person bring to the rescue center, in the case of multi-grasping.
- Fixed a bug in the calculation of the number of wounded person brought to the rescue center when they are brought by more than one drone.
- Spelling in README.md
- Compatibility with the new version of numpy : numpy.int has been deprecated, use int instead.

## [2.2.1] - 2022-12-15

### Changed
- Cleaning example.
- Disabled forbidden functions : position(), angle(), velocity(), angular_velocity() => generate an exception.

### Added
- Added map_intermediate_02.py !
- In README, added instructions to install the program on Windows10. 
- Added function grasped_entities() in drone_abstract.
- Added function measured_velocity() computed from odometer and compass values in drone_abstract.
- Added function measured_angular_velocity() in drone_abstract.
- Added some tests.

### Fixed
- bug fix : the execution of the code goes through your control() function even if you have activated the use of the keyboard to control the drone.
- there should be no more NaN at the first reading of the sensor data.
- installation bug.

## [2.2.0] - 2022-10-27

### Changed
- Refactoring of the map construction: now, drones are instantiated in the map class.

### Added
- Adding screen recorder tool: not really useful for competitors, but for evaluators. It is disabled by default.

### Fixed
- Fixing bug with macOS: the simulator crashed regularly on macOS. The problem is corrected but at the cost of a decrease in performance on macOS (some calculations done by the GPU are done by the CPU instead)

## [2.1.0] - 2022-10-21

### Changed
- change default GPS noise
- function name:
    - rename *measured_position()* to *measured_gps_position()* in the class *DroneAbstract*
    - rename *measured_angle()* to *measured_compass_angle()*  in the class *DroneAbstract*
 - the "no gps zone"  still disables the gps sensor and now, it disables also the compass sensor.

### Added
- add odometer sensor (see README.md for more information)
- remove velocity sensor (whose data can be replaced by a calculation on the odometry data)
- visualization of the noises on gps sensor, compass sensor and odometer sensor.

### Fixed
- improve code : typing, comments, documentation
- fix small bugs

## [2.0.0] - 2022-10-10

### Changed
- New version of simple-playgrounds. A lot of changes !!
- The interface is now more beautiful and faster.

## [1.5.2] - 2022-09-08

### Fixed
- Spelling correction in the README and comments of the code

## [1.5.1] - 2022-02-28

### Changed
- The radius, around the drones, of space that is considered explored, increases from 100 to 200

### Fixed
- Bug fix in the calculation of the explored area.

## [1.5.0] - 2022-02-15

### Changed
- *launcher.py* file is reformatted
- The evaluation report file is improved
- Information displayed in the console is improved

### Added

### Fixed
- The package *pathlib* has been changed to *pathlib2*, because *pathlib* is not maintained anymore.

## [1.4.1] - 2022-02-10

### Fixed
- Bug fix in *launcher.py* in function reset()

## [1.4.0] - 2022-02-04

### Changed
- *launcher.py* will now run the code following the competition evaluation rules (tests without special zones and with each zone activated)
- Special zones can now be activated/deactivated in the maps
- The starting points of the drones are not random anymore. The position of all drones now displays a square.

### Added
- New map : MyMapCompet02
- Teams should now fill the team_info.yml file in the solutions/ directory
- *launcher.py* now produces a pdf report of the performances in the *~/results_swarm_rescue/* directory according to the challenge evaluation rules

### Fixed


## [1.3.0] - 2022-01-27

### Changed
- Change fov of the *DroneLidar* from 180° to 360°, but we keep 90 rays.
- Split the sensor *DronePosition* in 2 sensors : *DroneGPS*, for only the position in pixels, and *DroneCompass*, for only the orientation of the drone in radians. Thus, the no-GPS zone will only apply to *DroneGP*S sensor, the orientation is still available.
- In the no-GPS zone, the measured position was always (0, 0). Now, it is (NaN, NaN).

### Added
- There are new functions in the class DroneAbstract to known if a sensor is disabled:
    - *touch_is_disabled()*
    - *semantic_cones_is_disabled()*

src/swarm_rescue/maps/walls_final_map.py
"""
This file was generated by the tool 'image_to_map.py' in the directory tools.
This tool permits to create this kind of file by providing it an image of the map we want to create.
"""

from spg_overlay.entities.normal_wall import NormalWall, NormalBox


# Dimension of the map : (1819, 1350)
# Dimension factor : 1.0
def add_boxes(playground):
    # box 0
    box = NormalBox(up_left_point=(-894, -508),
                    width=123, height=88)
    playground.add(box, box.wall_coordinates)

    # box 1
    box = NormalBox(up_left_point=(-694, -496),
                    width=80, height=164)
    playground.add(box, box.wall_coordinates)

    # box 2
    box = NormalBox(up_left_point=(-412, 305),
                    width=264, height=317)
    playground.add(box, box.wall_coordinates)

    # box 3
    box = NormalBox(up_left_point=(-384, 534),
                    width=118, height=120)
    playground.add(box, box.wall_coordinates)

    # box 4
    box = NormalBox(up_left_point=(618, 618),
                    width=89, height=148)
    playground.add(box, box.wall_coordinates)

    # box 5
    box = NormalBox(up_left_point=(72, 663),
                    width=89, height=157)
    playground.add(box, box.wall_coordinates)

    # box 6
    box = NormalBox(up_left_point=(-90, 663),
                    width=88, height=157)
    playground.add(box, box.wall_coordinates)

    # boxes cadre
    box = NormalBox(up_left_point=(-910, 675),
                    width=20, height=1350)
    playground.add(box, box.wall_coordinates)

    box = NormalBox(up_left_point=(-910, 675),
                    width=1820, height=16)
    playground.add(box, box.wall_coordinates)

    box = NormalBox(up_left_point=(-910, -655),
                    width=1820, height=20)
    playground.add(box, box.wall_coordinates)

    box = NormalBox(up_left_point=(887, 675),
                    width=23, height=1350)
    playground.add(box, box.wall_coordinates)

def add_walls(playground):
    # vertical wall 0
    wall = NormalWall(pos_start=(-891, 663),
                      pos_end=(-891, -656))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 1
    wall = NormalWall(pos_start=(-892, 661),
                      pos_end=(888, 661))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 2
    wall = NormalWall(pos_start=(-510, 661),
                      pos_end=(-510, 415))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 3
    wall = NormalWall(pos_start=(-152, 661),
                      pos_end=(-152, 612))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 4
    wall = NormalWall(pos_start=(-890, 659),
                      pos_end=(-86, 659))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 5
    wall = NormalWall(pos_start=(-8, 661),
                      pos_end=(-8, 510))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 6
    wall = NormalWall(pos_start=(-90, 512),
                      pos_end=(-8, 512))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 7
    wall = NormalWall(pos_start=(-8, 659),
                      pos_end=(76, 659))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 8
    wall = NormalWall(pos_start=(156, 661),
                      pos_end=(156, 510))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 9
    wall = NormalWall(pos_start=(74, 512),
                      pos_end=(156, 512))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 10
    wall = NormalWall(pos_start=(360, 661),
                      pos_end=(360, 325))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 11
    wall = NormalWall(pos_start=(156, 659),
                      pos_end=(886, 659))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 12
    wall = NormalWall(pos_start=(886, 661),
                      pos_end=(886, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 13
    wall = NormalWall(pos_start=(-512, 660),
                      pos_end=(-512, 415))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 14
    wall = NormalWall(pos_start=(-154, 660),
                      pos_end=(-154, 612))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 15
    wall = NormalWall(pos_start=(-88, 660),
                      pos_end=(-88, 511))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 16
    wall = NormalWall(pos_start=(74, 660),
                      pos_end=(74, 511))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 17
    wall = NormalWall(pos_start=(358, 660),
                      pos_end=(358, 325))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 18
    wall = NormalWall(pos_start=(884, 660),
                      pos_end=(884, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 19
    wall = NormalWall(pos_start=(621, 617),
                      pos_end=(621, 474))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 20
    wall = NormalWall(pos_start=(622, 616),
                      pos_end=(704, 616))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 21
    wall = NormalWall(pos_start=(702, 616),
                      pos_end=(702, 474))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 22
    wall = NormalWall(pos_start=(-890, 537),
                      pos_end=(-750, 537))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 23
    wall = NormalWall(pos_start=(-890, 535),
                      pos_end=(-752, 535))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 24
    wall = NormalWall(pos_start=(-382, 533),
                      pos_end=(-382, 418))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 25
    wall = NormalWall(pos_start=(-382, 532),
                      pos_end=(-270, 532))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 26
    wall = NormalWall(pos_start=(-272, 532),
                      pos_end=(-272, 418))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 27
    wall = NormalWall(pos_start=(-154, 516),
                      pos_end=(-154, 301))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 28
    wall = NormalWall(pos_start=(620, 476),
                      pos_end=(702, 476))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 29
    wall = NormalWall(pos_start=(-382, 420),
                      pos_end=(-272, 420))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 30
    wall = NormalWall(pos_start=(-702, 417),
                      pos_end=(-512, 417))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 31
    wall = NormalWall(pos_start=(-702, 415),
                      pos_end=(-510, 415))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 32
    wall = NormalWall(pos_start=(-890, 303),
                      pos_end=(-502, 303))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 33
    wall = NormalWall(pos_start=(-409, 304),
                      pos_end=(-409, -8))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 34
    wall = NormalWall(pos_start=(-408, 303),
                      pos_end=(-154, 303))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 35
    wall = NormalWall(pos_start=(-726, 303),
                      pos_end=(-726, 289))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 36
    wall = NormalWall(pos_start=(-890, 301),
                      pos_end=(-504, 301))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 37
    wall = NormalWall(pos_start=(-504, 303),
                      pos_end=(-504, -489))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 38
    wall = NormalWall(pos_start=(-153, 517),
                      pos_end=(-153, -7))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 39
    wall = NormalWall(pos_start=(-154, -5),
                      pos_end=(630, -5))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 40
    wall = NormalWall(pos_start=(-506, 302),
                      pos_end=(-506, -489))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 41
    wall = NormalWall(pos_start=(-726, 172),
                      pos_end=(-726, 21))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 42
    wall = NormalWall(pos_start=(-728, 171),
                      pos_end=(-728, 21))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 43
    wall = NormalWall(pos_start=(-890, 59),
                      pos_end=(-858, 59))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 44
    wall = NormalWall(pos_start=(-750, 59),
                      pos_end=(-634, 59))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 45
    wall = NormalWall(pos_start=(-524, 59),
                      pos_end=(-504, 59))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 46
    wall = NormalWall(pos_start=(-890, 57),
                      pos_end=(-858, 57))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 47
    wall = NormalWall(pos_start=(-752, 57),
                      pos_end=(-636, 57))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 48
    wall = NormalWall(pos_start=(-526, 57),
                      pos_end=(-504, 57))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 49
    wall = NormalWall(pos_start=(722, -5),
                      pos_end=(886, -5))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 50
    wall = NormalWall(pos_start=(-410, -7),
                      pos_end=(620, -7))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 51
    wall = NormalWall(pos_start=(730, -5),
                      pos_end=(730, -220))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 52
    wall = NormalWall(pos_start=(730, -7),
                      pos_end=(886, -7))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 53
    wall = NormalWall(pos_start=(728, -6),
                      pos_end=(728, -220))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 54
    wall = NormalWall(pos_start=(-726, -71),
                      pos_end=(-726, -194))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 55
    wall = NormalWall(pos_start=(-728, -72),
                      pos_end=(-728, -194))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 56
    wall = NormalWall(pos_start=(-412, -95),
                      pos_end=(-412, -655))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 57
    wall = NormalWall(pos_start=(-412, -97),
                      pos_end=(-292, -97))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 58
    wall = NormalWall(pos_start=(-412, -99),
                      pos_end=(-292, -99))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 59
    wall = NormalWall(pos_start=(-194, -99),
                      pos_end=(618, -99))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 60
    wall = NormalWall(pos_start=(-154, -98),
                      pos_end=(-154, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 61
    wall = NormalWall(pos_start=(170, -99),
                      pos_end=(170, -122))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 62
    wall = NormalWall(pos_start=(-196, -101),
                      pos_end=(620, -101))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 63
    wall = NormalWall(pos_start=(168, -100),
                      pos_end=(168, -122))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 64
    wall = NormalWall(pos_start=(-890, -192),
                      pos_end=(-858, -192))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 65
    wall = NormalWall(pos_start=(-754, -192),
                      pos_end=(-627, -192))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 66
    wall = NormalWall(pos_start=(-526, -192),
                      pos_end=(-504, -192))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 67
    wall = NormalWall(pos_start=(-890, 661),
                      pos_end=(-890, -512))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 68
    wall = NormalWall(pos_start=(-890, -194),
                      pos_end=(-860, -194))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 69
    wall = NormalWall(pos_start=(-670, -192),
                      pos_end=(-670, -408))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 70
    wall = NormalWall(pos_start=(-756, -194),
                      pos_end=(-628, -194))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 71
    wall = NormalWall(pos_start=(-526, -194),
                      pos_end=(-504, -194))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 72
    wall = NormalWall(pos_start=(-672, -193),
                      pos_end=(-672, -409))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 73
    wall = NormalWall(pos_start=(-314, -219),
                      pos_end=(-152, -219))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 74
    wall = NormalWall(pos_start=(-314, -221),
                      pos_end=(-152, -221))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 75
    wall = NormalWall(pos_start=(168, -226),
                      pos_end=(168, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 76
    wall = NormalWall(pos_start=(730, -315),
                      pos_end=(730, -336))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 77
    wall = NormalWall(pos_start=(728, -316),
                      pos_end=(728, -336))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 78
    wall = NormalWall(pos_start=(-412, -332),
                      pos_end=(-254, -332))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 79
    wall = NormalWall(pos_start=(-152, -332),
                      pos_end=(-106, -332))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 80
    wall = NormalWall(pos_start=(-10, -332),
                      pos_end=(215, -332))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 81
    wall = NormalWall(pos_start=(-410, -97),
                      pos_end=(-410, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 82
    wall = NormalWall(pos_start=(-338, -332),
                      pos_end=(-338, -554))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 83
    wall = NormalWall(pos_start=(-412, -334),
                      pos_end=(-254, -334))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 84
    wall = NormalWall(pos_start=(-152, -99),
                      pos_end=(-152, -655))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 85
    wall = NormalWall(pos_start=(-152, -334),
                      pos_end=(-106, -334))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 86
    wall = NormalWall(pos_start=(328, -334),
                      pos_end=(759, -334))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 87
    wall = NormalWall(pos_start=(852, -334),
                      pos_end=(886, -334))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 88
    wall = NormalWall(pos_start=(-340, -333),
                      pos_end=(-340, -554))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 89
    wall = NormalWall(pos_start=(169, -225),
                      pos_end=(169, -655))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 90
    wall = NormalWall(pos_start=(-10, -334),
                      pos_end=(214, -334))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 91
    wall = NormalWall(pos_start=(464, -334),
                      pos_end=(464, -518))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 92
    wall = NormalWall(pos_start=(326, -336),
                      pos_end=(758, -336))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 93
    wall = NormalWall(pos_start=(850, -336),
                      pos_end=(886, -336))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 94
    wall = NormalWall(pos_start=(462, -335),
                      pos_end=(462, -518))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 95
    wall = NormalWall(pos_start=(-670, -406),
                      pos_end=(-504, -406))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 96
    wall = NormalWall(pos_start=(-672, -408),
                      pos_end=(-504, -408))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 97
    wall = NormalWall(pos_start=(-252, -435),
                      pos_end=(-252, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 98
    wall = NormalWall(pos_start=(-254, -436),
                      pos_end=(-254, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 99
    wall = NormalWall(pos_start=(-691, -497),
                      pos_end=(-691, -655))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 100
    wall = NormalWall(pos_start=(-892, -653),
                      pos_end=(-692, -653))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 101
    wall = NormalWall(pos_start=(-690, -498),
                      pos_end=(-617, -498))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 102
    wall = NormalWall(pos_start=(-620, -498),
                      pos_end=(-620, -654))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 103
    wall = NormalWall(pos_start=(-890, -510),
                      pos_end=(-774, -510))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 104
    wall = NormalWall(pos_start=(-776, -510),
                      pos_end=(-776, -592))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 105
    wall = NormalWall(pos_start=(-889, -588),
                      pos_end=(-889, -654))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 106
    wall = NormalWall(pos_start=(-890, -590),
                      pos_end=(-776, -590))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 107
    wall = NormalWall(pos_start=(465, -620),
                      pos_end=(465, -655))
    playground.add(wall, wall.wall_coordinates)

    # vertical wall 108
    wall = NormalWall(pos_start=(462, -621),
                      pos_end=(462, -655))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 109
    wall = NormalWall(pos_start=(-620, -653),
                      pos_end=(886, -653))
    playground.add(wall, wall.wall_coordinates)

    # horizontal wall 110
    wall = NormalWall(pos_start=(-892, -655),
                      pos_end=(886, -655))
    playground.add(wall, wall.wall_coordinates)

